### PR TITLE
(proof-new) Add proofs for circuit propagator

### PR DESCRIPTION
**Pull request by nafur**
This PR adds code to generate proofs for inferences within the (non-clausal) circuit propagator.
It consists of many small methods that each implement simple derivations about single Boolean connectives. It uses already existing proof rules, and thus no separate proof checker is required.
A second PR will use these rules within the circuit propagator class.

### CODE REVIEW

File: src/theory/booleans/proof_circuit_propagator.h [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+  /** (not x) is true  -->  x is false (and vice versa) */
+  std::shared_ptr<ProofNode> Not(bool negate, Node parent);
+
+  /** (=> X false)  -->  (not X) */
+  std::shared_ptr<ProofNode> impliesXFromY(Node parent);
+  /** (=> true Y)  -->  Y */
+  std::shared_ptr<ProofNode> impliesYFromX(Node parent);
+
+  /** Derive X from (= X Y) */
+  std::shared_ptr<ProofNode> eqXFromY(bool y, Node Parent);
```
Comment (ajreynol) [inline src/theory/booleans/proof_circuit_propagator.h:62]: Don't capatilize Parent.

File: src/theory/booleans/proof_circuit_propagator.h [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+  /** (not x) is true  -->  x is false (and vice versa) */
+  std::shared_ptr<ProofNode> Not(bool negate, Node parent);
+
+  /** (=> X false)  -->  (not X) */
+  std::shared_ptr<ProofNode> impliesXFromY(Node parent);
+  /** (=> true Y)  -->  Y */
+  std::shared_ptr<ProofNode> impliesYFromX(Node parent);
+
+  /** Derive X from (= X Y) */
```
Comment (ajreynol) [inline src/theory/booleans/proof_circuit_propagator.h:61]: I assume this means from (= X Y) and Y?

File: src/theory/booleans/proof_circuit_propagator.h [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+  std::shared_ptr<ProofNode> eqXFromY(bool y, Node Parent);
+  /** Derive Y from (= X Y) */
```
Comment (ajreynol) [inline src/theory/booleans/proof_circuit_propagator.h:63]: Ditto.

File: src/theory/booleans/proof_circuit_propagator.h [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+  /** Evaluate (= X Y) from X,Y */
+  std::shared_ptr<ProofNode> eqEval(bool x, bool y);
+
+  /** Evaluate (=> X Y) from X,Y */
+  std::shared_ptr<ProofNode> impliesEval(bool premise, bool conclusion);
+  /** Evaluate (xor X Y) from X,Y */
+  std::shared_ptr<ProofNode> xorEval(bool x, bool y);
+
+ private:
+  Node d_child;
```
Comment (ajreynol) [inline src/theory/booleans/proof_circuit_propagator.h:188]: Document these.

File: src/theory/booleans/proof_circuit_propagator.h [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+   */
+  std::shared_ptr<ProofNode> iteIsCase(unsigned c);
+
+  /** (not (=> X Y))  -->  X */
+  std::shared_ptr<ProofNode> impliesNegX();
+  /** (not (=> X Y))  -->  (not Y) */
+  std::shared_ptr<ProofNode> impliesNegY();
+
+ private:
+  TNode d_parent;
```
Comment (ajreynol) [inline src/theory/booleans/proof_circuit_propagator.h:151]: Document these.

**Review by ajreynol — CHANGES_REQUESTED** [review]

File: src/theory/booleans/proof_circuit_propagator.h (hunk 2 above)
Comment (nafur) [reply src/theory/booleans/proof_circuit_propagator.h:61]: It means that `(= X Y)` holds and I derive the value of `X` from whatever `Y` is.

**Review by ajreynol — APPROVED** [review]
LGTM

**Comment by HanielB** [comment]
Marking as "do not merge" so that the instances of CHAIN_RESOLUTION are changed to RESOLUTION. That OK, @nafur ?

**Pull request merged**